ql/src/test/org/apache/hadoop/hive/ql/TestTxnCommands.java

```diff
@@ -378,6 +378,33 @@ private IMetaStoreClient prepareParallelTest(String tableName, int val)
     return msClient;
   }
 
+  @Test
+  public void testAddConstraintAdvancingWriteIds() throws Exception {
+
+    String tableName = "constraints_table";
+    hiveConf.setBoolean("hive.stats.autogather", true);
+    hiveConf.setBoolean("hive.stats.column.autogather", true);
+    // Need to close the thread local Hive object so that configuration change is reflected to HMS.
+    Hive.closeCurrent();
+    runStatementOnDriver("drop table if exists " + tableName);
+    runStatementOnDriver(String.format("create table %s (a int, b string) stored as orc " +
+        "TBLPROPERTIES ('transactional'='true', 'transactional_properties'='insert_only')",
+        tableName));
+    runStatementOnDriver(String.format("insert into %s (a) values (0)", tableName));
+    IMetaStoreClient msClient = new HiveMetaStoreClient(hiveConf);
+    String validWriteIds = msClient.getValidWriteIds("default." + tableName).toString();
+    LOG.info("ValidWriteIds before add constraint::"+ validWriteIds);
+    Assert.assertEquals("default.constraints_table:1:9223372036854775807::", validWriteIds);
+    runStatementOnDriver(String.format("alter table %s  ADD CONSTRAINT a_PK PRIMARY KEY (`a`) DISABLE NOVALIDATE", tableName));
+    validWriteIds  = msClient.getValidWriteIds("default." + tableName).toString();
+    LOG.info("ValidWriteIds after add constraint primary key::"+ validWriteIds);
+    Assert.assertEquals("default.constraints_table:2:9223372036854775807::", validWriteIds);
+    runStatementOnDriver(String.format("alter table %s CHANGE COLUMN b b STRING NOT NULL", tableName));
+    validWriteIds  = msClient.getValidWriteIds("default." + tableName).toString();
```

Comment:
Can we add the test for unique and check constraints as well for completeness?

Reply:
Added unique and check constraints. Please review.

Reply:
Thanks. Looks good!